KAFKA-7236: Add --under-min-isr option to describe topics command (KIP-351)

KevinLiLu opened the pull request:
[KIP-351](https://cwiki.apache.org/confluence/display/KAFKA/KIP-351%3A+Add+--under-min-isr+option+to+describe+topics+command)

- Add `--under-min-isr` option to `TopicCommand` to report partitions under the configured `min.insync.replicas` value

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

KevinLiLu committed:
KAFKA-7236: Add --under-min-isr option to describe topics command (KIP-351)

KevinLiLu committed:
Minor changes to description and make test consistent with others

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -99,29 +99,37 @@ object TopicCommand extends Logging {
                                    leader: Option[Int],
                                    assignedReplicas: Seq[Int],
                                    isr: Seq[Int],
+                                   minIsr: Int,
```
Since this is just a count, I suggest using `minIsrCount` to avoid confusion with `isr` which is also a list of `Int`.

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -220,13 +228,16 @@ object TopicCommand extends Logging {
           }
         }
         if (describeOptions.describePartitions) {
+          val computedMinIsr = if (opts.reportUnderMinIsrPartitions)
```
Similarly here: `configuredMinIsrCount`?

KevinLiLu commented:
@viktorsomogyi @gwenshap can you give this a review? Thanks~

vahidhashemian commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -536,6 +536,27 @@ class TopicCommandWithAdminClientTest extends KafkaServerTestHarness with Loggin
+  }
+
+  @Test
+  def testDescribeUnderMinIsrPartitions(): Unit = {
```
Could you also add at least one more unit tests, that verifies the output is empty if there is no under replicated partitions? Alternatively, you can add a unit test that verifies only valid partitions are listed when there is a mix of under-replicated and fully-replicated partitions.

KevinLiLu replied:
I've added an additional `testDescribeUnderMinIsrPartitionsMixed`.

vahidhashemian commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -536,6 +548,8 @@ object TopicCommand extends Logging {
                                                             "if set when describing topics, only show under replicated partitions")
     private val reportUnavailablePartitionsOpt = parser.accepts("unavailable-partitions",
                                                             "if set when describing topics, only show partitions whose leader is not available")
+    private val reportUnderMinIsrPartitionsOpt = parser.accepts("under-minisr-partitions",
```
In the KIP the config is proposed as `under-min-isr-partitions`. Is there a reason you used a different name? I think the name in the KIP (with a dash between `min` and `isr`) looks and reads better.

KevinLiLu replied:
Good catch, thanks!

vahidhashemian reviewed (commented):
Thank you for the KIP and the PR. Looks good to me overall. I left a few minor comments inline.

KevinLiLu committed:
Fix option, and add additional test with mixed partition status

KevinLiLu committed:
Add fully-replicated-topic to test case

KevinLiLu commented:
Hi @vahidhashemian, thanks for the review! I have added a commit addressing your comments.

I understand that the feature freeze for `2.2.0` was January 31st. Is this KIP bumped to the next release then, or can we still include this minor feature into `2.2.0`?

vahidhashemian commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -536,13 +536,78 @@ class TopicCommandWithAdminClientTest extends KafkaServerTestHarness with Loggin
       killBroker(0)
       val output = TestUtils.grabConsoleOutput(
         topicService.describeTopic(new TopicCommandOptions(Array("--under-replicated-partitions"))))
+      println(output)
```
nit: Unintentional `println` here?

vahidhashemian commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -536,13 +536,78 @@ class TopicCommandWithAdminClientTest extends KafkaServerTestHarness with Loggin
+    *   (4) topic with fully replicated partition
+    *
+    * Output should only display the (1) topic with partition under min ISR count and (3) topic with offline partition
+    */
```
nit: Please change references to three/3 to four/4 since there are 4 topics now.

vahidhashemian reviewed (commented):
Thanks for updating the PR. I left a couple of minor nits that you can hopefully fix quickly. Then, I'll approve the PR, and merge once the trunk version is bumped.

Unfortunately, since the cutoff for 2.2 is passed this will be merged to trunk for the following release.

KevinLiLu committed:
Address review nits

vahidhashemian reviewed (approved):
Thanks for PR. LGTM.

viktorsomogyi reviewed (approved):
Reviewed the PR, LGTM.

viktorsomogyi commented:
retest this please

KevinLiLu commented:
retest please

KevinLiLu commented:
retest this please

mimaison reviewed (approved):
LGTM

KevinLiLu commented:
Hi @vahidhashemian, thanks for the merge. I see that the trunk version was bumped up after this PR was merged. Did we squeeze this into `2.2.0` release, or is this still pushed back to `2.3.0`?

vahidhashemian commented:
@KevinLiLu I believe this KIP was originally included in 2.2.0 release plan. So this PR should also be included since the feature freeze is this Friday. @mjsax could you please confirm? Thanks!

mjsax commented:
KIPs must be merged before feature freeze deadline (cf. https://cwiki.apache.org/confluence/display/KAFKA/Time+Based+Release+Plan) that was two week ago. Thus, this KIP slips for 2.2. Sorry. 

It's already marked as postponed to 2.3 in https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=100827512

KevinLiLu commented:
@mjsax  no worries, thanks for the clarification!

vahidhashemian commented:
Thanks for clarifying @mjsax.
